proposals/0000-dependent-type-design.rst

```diff
@@ -0,0 +1,388 @@
+  to table the proposal until the features are ready, but also understood that we wouldn't
+  steal syntax invalidating `#102`_. That is, we implicitly refined criterion 2 to
+  include the syntax described in `#102`_, without directly committing to including
+  dependent types.
+
+* `#106`_: This accepted, unimplemented proposal describes a way to define a datatype
+  such that its constructors enter the type-level namespace, not the term-level namespace.
+  Some debate around the syntax worked hard to satisfy criteria 1 and 2, which was
+  acheived successfully. However, we may have settled upon different syntax without
+  having criterion 2 in mind.
+
+* `#236`_: This meta-proposal is another attempt to fill out details of criterion 2.
+  It has served as a useful place to imagine what dependent types in Haskell would
+  look like and to coordinate other proposals in fitting together.
+
+* `#242`_: This proposes unsaturated type family applications. A key challenge in accepting
+  unsaturated type families is in type inference: If we know ``a b ~ Maybe Int``, can
+  we conclude ``a ~ Maybe`` and ``b ~ Int``? Only if ``a`` is not a type family -- that
+  is, only if ``a`` is *matchable* (a combination of generative and injective).
+  (Section 4.2.4 of `my thesis`_ provides an introduction and should be understandable independent
+  of the rest of the thesis.) Matchability is properly the property of a function
+  arrow: we say that ``Maybe :: Type -> Type`` has a matchable arrow (because we
+  can match on ``Maybe Int`` in a type family to extract out the ``Int``) while
+  ``Id :: Type -> Type`` has an unmatchable arrow.
+
+  A key question is how we distinguish matchable arrows from unmatchable ones. Currently,
+  all arrows in types of types are matchable; all arrows in types of terms are unmatchbale.
+  Today, without dependent types, matchability only matters in the types of types because
+  matchability really is needed only to inform type inference. (We don't yet perform
+  *term* inference.) Conversely, `linear types`_ matter only in the types of terms;
+  we don't yet have compile-time linearity. So, we might imagine using the same
+  syntax for both linear types as for matchability. In practice, without dependent types,
+  there would be no conflict. Yet if we are exploring dependent types, such a
+  syntax would be terribly forward-incompatible.
+
+  As it turns out, there is enough syntactic space for these two features to avoid
+  each other (and thus satisfy both criteria 1 and 2), but this choice had to be
+  made intentionally.
+
+  A separate question is one of defaults: when we write ``Type -> Type``, should
+  that arrow be matchable or unmatchable? The proposal describes the choice here
+  as a tension between backward compatibility and forward compatibility. (To be
+  fair, though, there isn't a true backward-compatibility problem, as the matter
```

Comment:
It looks like I have a slightly different understanding of what backward-compatibility means. Sure, introducing a feature that breaks programs that do not use the relevant pragma is concerning, but the way I think of compatibility is "does enabling this pragma in an existing source file change the behaviour of the program either at runtime, or at compile time?"

In that sense, I consider `PolyKinds` to be backward incompatible, but `TypeApplications` to be perfectly compatible (maybe modulo some syntax stealing?). Granted, this notion of compatibility is hard to exactly characterise, because some extensions are "more compatible" than others. E.g. `ExplicitForalls` is much less likely to cause breakage than `ScopedTypeVariables`.

In this sense, the alternatives are between an option where you can turn on the extension and your existing programs still compile, and an option where you might need to change some types, but the result is more ergonomic moving forward.

Maybe this point at a need to more carefully characterise what level of compatibility breakage we're happy to tolerate. Is the rule that we are careful to not to change any program that doesn't turn on the extension, but anything else is fair game?

Reply:
Hm -- I actually think you and I have the same understanding of backward-compatibility here. That is, I would like backward-compatibility to mean that enabling a new extension keeps all my code working just fine (that is, with the same meaning). This is why I distinguish between "regular" backward-compatibility and "true" backward-compatibility.

Regarding your question: I think we should indeed be quite careful (as much as possible -- I'm still worried about `forall` becoming a keyword, which may be unavoidable) about not breaking code that doesn't enable an extension. For code that does enable an extension, any breakage is unfortunate and should be considered a downside of a proposal. Yet, I would say that it's possible that upsides of the proposal may outweigh that downside. It all depends on the proposal.